### ISSUE TITLE
Bookmarks/Clips

### ISSUE THREAD

**Issue by gabriel-vanca** [rmcrackan/Libation#345]
Would it be possible to make bookmarks/clips available? Perhaps they can be downloaded too when the book is downloaded.

**Comment by Mbucari** [comment]
I think it's possible to download bookmarks and clips through the API, but the question is what file format would they be saved in, and how would you use them? Bookmarks and clips are designed to work with the Audible audiobook player. I suppose clips could be just that, pieces of the audiobooks stored as individual mp4 or mp3. But I don't know about bookmarks. How would you like to see bookmarks saved to a file?

**Comment by Mbucari** [comment]
Bookmark timestamps and annotation text are easy: just put them in a CSV which you can then import into pretty much any software ever created. For the accompanying audio file, maybe:
1. Save the 3-ish second clip into same folders as a CSV
2. At a third column in the CSV with the audio files name

Sound good?

**Comment by Mbucari** [comment]
Going back to the clips, I'm not crazy about having to split chunks of the audio out into their own files. Starting with what we have right now, that would just be a lot of changes to the audio processing routines. It would be a lot of work and I don't have confidence I'd get it right on the first try.

So instead of that, what about .cue files? One .cue file per clip, or all clips in a single Bookmarks.cue file?

In fact, bookmarks could be done in pretty much the same way. One .cue file for all of your bookmarks.

**Comment by rmcrackan** [comment]
I don't use the bookmark features myself and I don't know how audible stores them. It sounds like you want small portions of the main audio file. Depending on how they're stored in your account, sounds to me like the output of a book + some time markers (which might be manageable by a cue file). Am I close? If not, where am I off?

**Comment by rmcrackan** [comment]
@mkb79 What do you know about bookmarks? I see [this](https://audible.readthedocs.io/en/latest/misc/external_api.html#get-https---cde-ta-g7g.amazon.com-FionaCDEServiceEngine-sidecar). However, that doesn't look like bookmarks from an account.

**Comment by mkb79** [comment]
@rmcrackan This will give you the bookmarks, clips and annotations for the given asin and the authenticated user. [Here](https://github.com/mkb79/audible-cli/blob/94e2d9a713973393c2c13d3802e7c82b92e6e752/src/audible_cli/models.py#L334) is the implementation for audible-cli.

**Comment by Mbucari** [comment]
So I've been trying to figure this out with help from @mkb79 [over here](https://github.com/mkb79/Audible/discussions/113), and it's frustrating. It appears that the Audible app doesn't sync clips and bookmarks to the server reliably. Additionally, I can't seem to retrieve clips from the api. I can set them in the api and they show up on my phone, but I can't query them.

**Comment by Mbucari** [comment]
@rmcrackan leave it for now. My obsessive personality won't allow me to give up this quickly, and @mkb79 says he'll try with his US account tomorrow.

**Comment by Mbucari** [comment]
@rmcrackan We might have this figured out. My existing auth tokens didn't grant me access to view clips. I registered a new device, and the new token showed me clips. @mkb79 says this is a relatively new feature and old auth files may not have access to clips. Can you do me a favor and try to retrieve clips from an audiobook on your account using [audible-cli](https://github.com/mkb79/audible-cli) (if your COVID brain will allow)? You'll need to create a clip for an audiobook in the mobile app, then run this command
```Console
audible api -i 4 -p key=[ASIN] -p type=AUDI https://cde-ta-g7g.amazon.com/FionaCDEServiceEngine/sidecar
```

You should get results like this:

```Json
{
    "md5": "XXX",
    "payload": {
        "acr": null,
        "guid": "_LATEST_",
        "key": "[ASIN]",
        "type": "AUDI",
        "records": [
            {
                "lastModificationTime": "2023-01-02 21:05:01.0",
                "annotationId": "XXX",
                "startPosition": "2000000",
                "type": "audible.clip",
                "endPosition": "2025000",
                "creationTime": "2023-01-02 21:05:01.0",
                "metadata": {
                    "note": "Test Clip notice",
                    "c_version": "10748071"
                }
            },
            {
                "lastModificationTime": "2023-01-02 21:05:02.0",
                "annotationId": "XXX",
                "startPosition": "2000000",
                "type": "audible.bookmark",
                "creationTime": "2023-01-02 21:05:02.0"
            },
            {
                "startPosition": "2025000",
                "type": "audible.last_heard",
                "creationTime": "2023-01-04 03:03:33.0"
            }
        ]
    }
}

```
If you see a record with `"type": "audible.clip"`, your token has access to that data.

**Comment by rmcrackan** [comment]
@Mbucari 

```
>audible api -i 4 -p key=B084RPCT73 -p type=AUDI https://cde-ta-g7g.amazon.com/FionaCDEServiceEngine/sidecar
{
    "md5": "[redacted]",
    "payload": {
        "key": "[redacted -- 10 chars]",
        "guid": "_LATEST_",
        "type": "AUDI",
        "acr": null,
        "records": [
            {
                "creationTime": "2023-01-04 20:15:44.0",
                "type": "audible.last_heard",
                "startPosition": "10018042"
            },
            {
                "endPosition": "10019671",
                "creationTime": "2023-01-04 20:15:44.0",
                "metadata": {
                    "c_version": "13997221"
                },
                "lastModificationTime": "2023-01-04 20:15:44.0",
                "type": "audible.clip",
                "startPosition": "9981942",
                "annotationId": "[redacted -- 14 chars]"
            },
            {
                "creationTime": "2023-01-04 20:15:45.0",
                "lastModificationTime": "2023-01-04 20:15:45.0",
                "type": "audible.bookmark",
                "startPosition": "9981942",
                "annotationId": "[redacted -- 14 chars]"
            }
        ]
    }
}
```

**Comment by Mbucari** [comment]
Nice! Now that I know the problem was just with my auth, this feature is nearly complete in the AudibleApi lib. Then it's a matter of figuring out what the Libation feature will look like. Let's have a race: can I complete this feature before you and your family recover lol

**Comment by rmcrackan** [comment]
You have a head start. Me and my wife were symptomatic on Jan 1 (happy new year!) and all kids started yesterday, thus resetting the clock

**Comment by Mbucari** [comment]
All right @gabriel-vanca , here's what I've got.

Every book has a new menu item:
![MenuItem](https://user-images.githubusercontent.com/37587114/210947414-4eebb7d2-c3ad-41bf-b7d6-0faf6382df17.png)

Which opens a new window:

![Window](https://user-images.githubusercontent.com/37587114/210947447-0151bc71-06a2-40ca-a23c-2719e0b636d5.png)

And lets yoy delete records and export them to xklsx, csv, and json.


<details>
  <summary>Here's the CSV</summary>

  ```CSV
  Type,Created,Start,AnnotationId,LastModified,End,Text,Title
  Clip,1/5/2023 11:50:47 PM -07:00,01:23:20,XXX,1/5/2023 11:50:47 PM -07:00,01:23:50,clip 1 note,clip 1 title
  Clip,1/5/2023 11:50:47 PM -07:00,01:40:00,XXX,1/5/2023 11:50:47 PM -07:00,01:40:30,clip 3 note,clip 3 title
  Note,1/5/2023 11:50:47 PM -07:00,11:40:00,XXX,1/5/2023 11:50:47 PM -07:00,11:43:20,Standalone note 2
  Note,1/5/2023 11:50:47 PM -07:00,01:23:20,XXX,1/5/2023 11:50:47 PM -07:00,01:23:20,clip 1 note
  Note,1/5/2023 11:50:47 PM -07:00,01:40:00,XXX,1/5/2023 11:50:47 PM -07:00,01:40:00,clip 3 note
  Note,1/5/2023 11:50:47 PM -07:00,19:10:00,XXX,1/5/2023 11:50:47 PM -07:00,19:25:00,Standalone note 1
  Bookmark,1/5/2023 11:50:47 PM -07:00,01:23:20,XXX,1/5/2023 11:50:47 PM -07:00
  Bookmark,1/5/2023 11:50:47 PM -07:00,01:40:00,XXX,1/5/2023 11:50:47 PM -07:00
  Bookmark,1/5/2023 11:50:47 PM -07:00,16:40:00,XXX,1/5/2023 11:50:47 PM -07:00
  Bookmark,1/5/2023 11:50:47 PM -07:00,19:26:40,XXX,1/5/2023 11:50:47 PM -07:00
  LastHeard,1/5/2023 11:50:47 PM -07:00,02:07:34.3210000
  ```  
</details>



<details>
  <summary>And here's the Json</summary>

  ```Json
{
    "title": "Harry Potter and the Half-Blood Prince, Book 6",
    "asin": "B017V4NOZ0",
    "exportTime": "2023-01-05T23:57:06.2021725-07:00",
    "records": [
      {
        "Type": "Note",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "11:40:00",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00",
        "End": "11:43:20",
        "Text": "Standalone note 2"
      },
      {
        "Type": "Bookmark",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "01:23:20",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00"
      },
      {
        "Type": "Note",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "01:23:20",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00",
        "End": "01:23:20",
        "Text": "clip 1 note"
      },
      {
        "Type": "Clip",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "01:23:20",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00",
        "End": "01:23:50",
        "Text": "clip 1 note",
        "Title": "clip 1 title"
      },
      {
        "Type": "Clip",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "01:40:00",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00",
        "End": "01:40:30",
        "Text": "clip 3 note",
        "Title": "clip 3 title"
      },
      {
        "Type": "Bookmark",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "01:40:00",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00"
      },
      {
        "Type": "Note",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "01:40:00",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00",
        "End": "01:40:00",
        "Text": "clip 3 note"
      },
      {
        "Type": "Bookmark",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "16:40:00",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00"
      },
      {
        "Type": "Note",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "19:10:00",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00",
        "End": "19:25:00",
        "Text": "Standalone note 1"
      },
      {
        "Type": "Bookmark",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "19:26:40",
        "AnnotationId": "XXX",
        "LastModified": "2023-01-05T23:50:47-07:00"
      },
      {
        "Type": "LastHeard",
        "Created": "2023-01-05T23:50:47-07:00",
        "Start": "02:07:34.3210000"
      }
    ]
  }
  ```  
</details>

I plan to add a cli command for these as well (similar to the export library command). You said you also wanted a setting to download this data with the books too, right?

**Comment by rmcrackan** [comment]
Included in latest: [v8.8.1](https://github.com/rmcrackan/Libation/releases/latest)
